src/main/java/co/cask/gcp/speech/SpeechToTextTransform.java

```diff
@@ -0,0 +1,377 @@
+import com.google.protobuf.ByteString;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.ws.rs.Path;
+
+/**
+ * Google Cloud Speech-To-Text Transform which transcripts audio
+ */
+@Plugin(type = Transform.PLUGIN_TYPE)
+@Name(SpeechToTextTransform.NAME)
+@Description(SpeechToTextTransform.DESCRIPTION)
+public class SpeechToTextTransform extends Transform<StructuredRecord, StructuredRecord> {
+  public static final String NAME = "SpeechToText";
+  public static final String DESCRIPTION = "Converts audio files to text by applying powerful neural network models.";
+  private SpeechTransformConfig config;
+  private Schema outputSchema = null;
+  private RecognitionConfig recognitionConfig;
+  private SpeechClient speech;
+
+  @Override
+  public void configurePipeline(PipelineConfigurer configurer) throws IllegalArgumentException {
+    super.configurePipeline(configurer);
+    config.validate(configurer.getStageConfigurer().getInputSchema());
+  }
+
+  @Override
+  public void initialize(TransformContext context) throws Exception {
+    super.initialize(context);
+    config.validate(context.getInputSchema());
+    if (context.getInputSchema() != null) {
+      outputSchema = getSchema(context.getInputSchema(), config.getPartsField(), config.getTextField());
+    }
+    setRecognitionConfig();
+    speech = SpeechClient.create(getSettings());
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) {
+    ByteString audioBytes = ByteString.copyFrom((byte[]) input.get(config.audioField));
+    RecognitionAudio audio = RecognitionAudio.newBuilder()
+      .setContent(audioBytes)
+      .build();
+
+    RecognizeResponse response = speech.recognize(recognitionConfig, audio);
+    List<SpeechRecognitionResult> results = response.getResultsList();
+
+    // if an output schema is available then use it or else use the schema for the given input record
+    Schema currentSchema;
+    if (outputSchema != null) {
+      currentSchema = outputSchema;
+    } else {
+      currentSchema = getSchema(input.getSchema(), config.getPartsField(), config.getTextField());
+    }
+
+    StructuredRecord.Builder outputBuilder = StructuredRecord.builder(currentSchema);
+
+    List<StructuredRecord> transcriptParts = new ArrayList<>();
+    StringBuilder completeTranscript = new StringBuilder();
+
+    Schema partsSchema = null;
+    if (config.getPartsField() != null && !config.getPartsField().isEmpty()) {
+      partsSchema = currentSchema.getField(config.getPartsField()).getSchema().getComponentSchema();
+    }
+
+    for (SpeechRecognitionResult result : results) {
+      if (config.getPartsField() != null && !config.getPartsField().isEmpty()) {
+        addTranscriptWithConfidence(partsSchema, transcriptParts, result.getAlternativesList());
```

Comment:
Did it in this way since outputSchema might not be known in all cases.

Reply:
the output schema might not be known, but the schema for the parts field will always be known. It should be created at initialize time. Instead of getting it from the outputschema, the getSchema() method should use the field's schema.